src/main/java/com/yelp/nrtsearch/server/luceneserver/search/collectors/MyTopSuggestDocsCollector.java

```diff
@@ -79,7 +93,119 @@ public TopSuggestDocs reduce(Collection<TopSuggestDocsCollector> collectors)
+              .sorted((o1, o2) -> -1 * Float.compare(o1.score, o2.score)) // highest score first
+              .limit(this.numHitsToCollect)
+              .toArray(TopSuggestDocs.SuggestScoreDoc[]::new);
+
+      return new TopSuggestDocs(
+          new TotalHits(this.numHitsToCollect, TotalHits.Relation.EQUAL_TO), suggestScoreDocs);
+    }
+  }
+
+  public static class ExhaustiveTopSuggestDocsCollector extends TopSuggestDocsCollector {
+    private final SuggestScoreDocPriorityQueue priorityQueue;
+
+    public ExhaustiveTopSuggestDocsCollector(int num) {
+      super(num, false);
+      priorityQueue = new SuggestScoreDocPriorityQueue(num);
+    }
+
+    @Override
+    protected void doSetNextReader(LeafReaderContext context) {
+      docBase = context.docBase;
+      // clears priority queue
+      priorityQueue.getResults();
+    }
+
+    @Override
+    public void collect(int docID, CharSequence key, CharSequence context, float score)
+        throws IOException {
+      TopSuggestDocs.SuggestScoreDoc suggestScoreDoc =
+          new TopSuggestDocs.SuggestScoreDoc(docBase + docID, key, context, score);
+      priorityQueue.add(suggestScoreDoc);
+    }
+
+    @Override
+    public TopSuggestDocs get() throws IOException {
+      TopSuggestDocs.SuggestScoreDoc[] results = priorityQueue.getResults();
+      if (results.length > 0) {
+        return new TopSuggestDocs(
+            new TotalHits(results.length, TotalHits.Relation.EQUAL_TO), results);
+      } else {
+        return TopSuggestDocs.EMPTY;
+      }
+    }
+  }
+
+  public static class SuggestScoreDocPriorityQueue {
+    private final Map<Integer, TopSuggestDocs.SuggestScoreDoc> docIdsMap;
+    private final PriorityQueue<TopSuggestDocs.SuggestScoreDoc> priorityQueue;
+
+    public SuggestScoreDocPriorityQueue(int num) {
+      docIdsMap = new HashMap<>(num);
+      priorityQueue = new PriorityQueue<>(num, (o1, o2) -> -1 * Float.compare(o1.score, o2.score));
+    }
+
+    /**
+     * Adds the suggestScoreDoc to the queue on the condition that: 1. SuggestScoreDoc with the same
+     * docId and greater score has not been collected before 2. The suggestScoreDoc's score is
+     * higher than the lowest score previously collected
+     *
+     * @param suggestScoreDoc document to add to the queue
+     * @return `true` if the operation changed the state of the queue, `false` otherwise
+     */
+    public boolean add(TopSuggestDocs.SuggestScoreDoc suggestScoreDoc) {
+      if (!docIdsMap.containsKey(suggestScoreDoc.doc)
+          || docIdsMap.get(suggestScoreDoc.doc).score < suggestScoreDoc.score) {
+        TopSuggestDocs.SuggestScoreDoc previousValue =
+            docIdsMap.put(suggestScoreDoc.doc, suggestScoreDoc);
+        if (previousValue != null) {
+          // remove old value from priority queue
+          priorityQueue.remove(previousValue);
+        }
+        // add new value to priority queue
+        priorityQueue.add(suggestScoreDoc);
+      }
```

Comment:
You will need see how performant this is. Some of the operations are O(n) with queue size.

I suspect that ranking will process the hits sequentially by doc id. If that is confirmed, you only need to keep track of the last doc id and score, and defer adding to the queue until you see a new id. This may also be possible to do at the collector level, so you can use the standard `SuggestScoreDocPriorityQueue`.

Reply:
> I suspect that ranking will process the hits sequentially by doc id. 

I tested this and it is not sequential by doc id.

---

In terms of performance - the slowest operation here `O(n)` is on `priorityQueue.remove`. I have not given it much thought on performance as I said [here](https://github.com/Yelp/nrtsearch/pull/499#issuecomment-1241737156). We can test this once we get more data and see how it performs.